File: c/src/audio.c
/**
 * @file audio.c
 * @brief Audio processing functionality for waveform generation
 */

// MINIMP3_IMPLEMENTATION and MINIMP3_NO_SIMD are defined in the Makefile

#include "waver.h"
#include "minimp3.h"
#include "minimp3_ex.h"

#include <stdio.h>
#include <string.h>
#include <stdint.h>

// Define INT16_MIN if not defined
#ifndef INT16_MIN
#define INT16_MIN (-32768)
#endif

/**
 * @brief Absolute value of a PCM sample
 *
 * This function converts a PCM sample into its positive only value
 * It handles the special case of sample being INT16_MIN value.
 *
 * @param sample The signed PCM sample
 * @return The absolue value of the signed PCM sample
 */
static unsigned int abs_pcm(int16_t sample) {
    if (sample == INT16_MIN) {
        return 32767;
    }
    return (sample < 0) ? -sample : sample;
}

/**
 * @brief Process an audio file and generate a waveform visualization
 *
 * This function decodes the audio file frame by frame and immediately processes
 * each frame to generate the waveform image, without storing all audio data in memory.
 *
 * @param input_path Path to the input audio file
 * @param image The waveform image to draw into
 * @param width Width of the output image in pixels
 * @return true if successful, false otherwise
 */
static bool process_audio_file(const char *input_path, waver_image_t *image, unsigned int width) {
    mp3dec_ex_t mp3d;
    if (mp3dec_ex_open(&mp3d, input_path, MP3D_SEEK_TO_SAMPLE)) {
        return false;
    }

    // Get channel information
    unsigned int channel_count = mp3d.info.channels;
    if (channel_count > 2) {
        channel_count = 2; // Limit to stereo
    }

    // Get total number of frames for scaling calculation
    uint64_t total_samples = mp3d.samples / mp3d.info.channels;
    if (total_samples == 0) {
        mp3dec_ex_close(&mp3d);
        return false;
    }

    // Calculate samples per pixel ratio for even distribution
    uint64_t samples_per_pixel = total_samples / width;
    uint64_t remainder = total_samples % width;

    // Use a reasonably sized buffer for streaming
    const size_t BUFFER_SIZE = 4096;
    mp3d_sample_t pcm[BUFFER_SIZE];

    // Initialize tracking variables
    uint64_t sample_count = samples_per_pixel;
    uint64_t sample_remainder = 0;

    unsigned int current_pixel = 0;  // The pixel we're currently collecting samples for
    unsigned int left = 0;
    unsigned int right = 0;


    // Process audio stream packet by packet
    size_t samples_read;
    while ((samples_read = mp3dec_ex_read(&mp3d, pcm, BUFFER_SIZE)) > 0) {
        // Process each frame in the current packet
        for (size_t i = 0; i < samples_read; i += channel_count) {
            // Process each channel's sample
            unsigned int sample = abs_pcm(pcm[i]);
            if (sample > left) {
                left = sample;
            }

            if (channel_count > 1) {
                sample = abs_pcm(pcm[i + 1]);
                if (sample > right) {
                    right = sample;
                }
            }

            sample_count--;
            if (sample_count == 0) {
                // Draw the current pixel with its max values
                if (channel_count > 1) {
                    waver_image_draw_point(image, current_pixel, left, right);
                } else {
                    waver_image_draw_point_mono(image, current_pixel, left);
                }

                // Reset max values for the new pixel
                left = 0;
                right = 0;

                // Update current pixel
                current_pixel++;
                sample_count = samples_per_pixel;
                sample_remainder += remainder;
                if (sample_remainder >= width) {
                    sample_remainder -= width;
                    sample_count++;
                }
            }
        }
    }

    // Draw the final pixel if we've collected any samples for it
    if (sample_count > 0) {
        if (channel_count > 1) {
            waver_image_draw_point(image, current_pixel, left, right);
        } else {
            waver_image_draw_point_mono(image, current_pixel, left);
        }
    }

    mp3dec_ex_close(&mp3d);
    return true;
}

/**
 * @brief Generate a waveform from an audio file
 *
 * @param input_path Path to the input audio file
 * @param output_path Path to save the output PNG
 * @param args Command-line arguments
 * @return true if successful, false otherwise
 */
bool waver_generate_waveform(
    const char *input_path,
    const char *output_path,
    const waver_args_t *args
) {
    if (!input_path || !output_path || !args) {
        return false;
    }

    // Skip if output exists and overwrite isn't allowed
    if (!args->overwrite) {
        FILE *test_file = fopen(output_path, "rb");
        if (test_file) {
            fclose(test_file);
            if (args->verbose) {
                waver_print_stderr(args, "Output file '%s' already exists - use --overwrite", output_path);
            }
            return true; // Not an error, but skipped
        }
    }

    // Generate the image buffer
    waver_image_t *image = waver_image_new(args->width, args->height);
    if (!image) {
        waver_print_stderr(args, "Failed to create image buffer");
        return false;
    }

    // Process audio file and generate waveform
    bool success = process_audio_file(input_path, image, args->width);
    if (!success) {
        waver_print_stderr(args, "Failed to process audio file: %s", input_path);
        waver_image_free(image);
        return false;
    }

    // Save or log the result
    if (!args->dry_run) {
        success = waver_image_save_optimized_png(
            image,
            &args->bg_color,
            &args->left_color,
            &args->right_color,
            output_path
        );

        if (success) {
            waver_print_stdout(args, "Created %s", output_path);
        } else {
            waver_print_stderr(args, "Failed to save PNG file: %s", output_path);
        }
    } else if (args->verbose) {
        waver_print_verbose(args, "DryRun %s", output_path);
        success = true;
    }

    waver_image_free(image);
    return success;
}

File: rust/src/audio.rs
/// Audio processing functionality for waveform generation.
use std::fs::File;
use std::path::Path;

use symphonia::core::audio::{AudioBuffer, Signal};
use symphonia::core::codecs::{DecoderOptions, CODEC_TYPE_NULL};
use symphonia::core::formats::FormatOptions;
use symphonia::core::io::MediaSourceStream;
use symphonia::core::meta::MetadataOptions;
use symphonia::core::probe::Hint;

use crate::cli::{AudioPath, WaverArgs};
use crate::error::{Result, WaverError};
use crate::image::WaveImage;

/// Generates a waveform visualization from an audio file.
///
/// # Arguments
///
/// * `input_path` - Path to the input audio file
/// * `output_path` - Path where the output PNG will be saved
/// * `args` - Command-line arguments containing configuration
///
/// # Returns
///
/// `Ok(())` on success, or an error if processing fails
pub fn generate_waveform(
    input_path: &AudioPath,
    output_path: impl AsRef<Path>,
    args: &WaverArgs,
) -> Result<()> {
    let input_path = input_path.path();
    let output_path = output_path.as_ref();

    // Skip if output exists and overwrite isn't allowed
    if !args.overwrite && output_path.exists() {
        if args.verbose {
            return Err(WaverError::generation_error(format!(
                "Output file '{}' already exists - use --overwrite",
                output_path.display()
            )));
        }
        return Ok(());
    }

    // Generate the image buffer
    let mut image = WaveImage::new(args.width, args.height);

    // Process audio file and generate waveform
    process_audio_file(input_path, &mut image, args.width())?;

    // Save or log the result
    if !args.dry_run {
        image.save_png(&args.background_color, &args.left_color, &args.right_color, output_path)?;
        args.print_to_stdout(&format!("Created {}", output_path.display()));
    } else if args.verbose {
        args.print_verbose(&format!("DryRun {}", output_path.display()));
    }

    Ok(())
}

/// Processes an audio file and generates a waveform visualization using a streaming approach.
///
/// This function opens an audio file, decodes it frame by frame, and immediately
/// processes each frame to generate the waveform image, without storing all audio data in memory.
///
/// # Performance
///
/// This is a performance-critical function.  It uses a streaming approach rather than buffering the
/// entire audio file, which results in:
/// - ~24x lower memory usage
/// - ~6.7x faster execution time
/// - Significantly fewer system calls
///
/// Do not change this to buffer all audio samples, as that would cause severe performance degradation.
///
/// # Arguments
///
/// * `input_path` - Path to the input audio file
/// * `image` - The waveform image to draw into
/// * `width` - Width of the output image in pixels
///
/// # Returns
///
/// `Ok(())` on success, or an error if processing fails
fn process_audio_file(input_path: &Path, image: &mut WaveImage, width: u32) -> Result<()> {
    // Open and probe the audio file
    let file = File::open(input_path)?;
    let mss = MediaSourceStream::new(Box::new(file), Default::default());

    let probed = symphonia::default::get_probe().format(
        &Hint::new(),
        mss,
        &FormatOptions::default(),
        &MetadataOptions::default(),
    )?;

    // Extract the first audio track
    let mut format = probed.format;
    let track = format
        .tracks()
        .iter()
        .find(|t| t.codec_params.codec != CODEC_TYPE_NULL)
        .ok_or_else(|| WaverError::generation_error(format!(
            "No audio track found in '{}'",
            input_path.display()
        )))?;

    // Initialize decoder
    let mut decoder = symphonia::default::get_codecs().make(
        &track.codec_params,
        &DecoderOptions { ..Default::default() },
    )?;

    // Get channel information
    let channel_count = track
        .codec_params
        .channels
        .map(|c| c.count())
        .unwrap_or(1)
        .min(2) as usize;

    // Get total number of frames (samples per channel) for scaling calculation
    let total_samples = track.codec_params.n_frames.unwrap_or(0).max(1) as u64;

    // Calculate samples per pixel and the fractional
    // samples per pixel in 1/width units - since we have
    // to use width as u64 a number of times, do that conversion once
    let width64 = width as u64;
    let samples_per_pixel = total_samples / width64;
    let fractional_samples = total_samples % width64;

    // Initialize state variables for processing
    let mut left = 0.0f32;
    let mut right = 0.0f32;
    let mut sample_progress = samples_per_pixel;
    let mut partial_progress = 0 as u64;
    let mut pixel_pos = 0;

    // Process audio stream packet by packet
    while let Ok(packet) = format.next_packet() {
        let decoded = decoder.decode(&packet)?;
        let mut buffer = AudioBuffer::<f32>::new(decoded.capacity() as u64, *decoded.spec());
        decoded.convert(&mut buffer);

        // Process each frame in the current packet
        for frame in 0..buffer.frames() {
            // Update max amplitude values for each channel
            left = left.max(buffer.chan(0)[frame].abs().min(1.0));
            if channel_count > 1 {
                right = right.max(buffer.chan(1)[frame].abs().min(1.0));
            }

            // Map samples to pixels
            sample_progress -= 1;

            if sample_progress == 0 {
                // When we've accumulated enough samples for a pixel, draw it
                if channel_count > 1 {
                    image.draw_point(pixel_pos, left, right);
                } else {
                    image.draw_point_mono(pixel_pos, left);
                }
                left = 0.0;
                right = 0.0; // Reset max values for next pixel
                pixel_pos += 1;
                sample_progress = samples_per_pixel;
                partial_progress += fractional_samples;
                // If we got enough fractional samples to get another
                // sample in this next section, bump it by one and
                // subtract the width.
                if partial_progress >= width64 {
                    partial_progress -= width64;
                    sample_progress += 1;
                }
            }
        }
    }

    // Draw any remaining partial pixel
    if pixel_pos < width {
        if channel_count > 1 {
            image.draw_point(pixel_pos, left, right);
        } else {
            image.draw_point_mono(pixel_pos, left);
        }
    }

    Ok(())
}